### PR TITLE
GenBCode: cleanup optimizations

### PR DESCRIPTION
**Pull request by lrytz**
Method-level cleanup optimizations.

See individual commit messages.

### CODE REVIEW

**Comment by soc** [comment]
Mh, just found a few weirdnesses and regressions with exceptions in GenBCode compared to GenASM. I'll wait until your improvements are merged, test again and if they still exist file tickets for them.

**Comment by lrytz** [comment]
If you have examples at hand, I'd prefer to have them earlier, just file a bug or send them to me.

**Comment by soc** [comment]
Done: https://issues.scala-lang.org/browse/SI-8957
Looks like all the other issues I found in 2.11.2 have been fixed somewhere between 2.11.2 and 2.11.4, just updated my REPL, checked and they are all gone. :-)

**Comment by lrytz** [comment]
> the other issues I found in 2.11.2 have been fixed somewhere between 2.11.2 and 2.11.4

Might be only due to dead-code elimination, did you try with `-Yopt:l:none`?

**Comment by retronym** [comment]
PLS REBUILD/pr-scala@d39525ac06c8d319b43d3388d624fb2c1dcbd601

**Comment by retronym** [comment]
Great work, @lrytz / @magarciaEPFL !

**Comment by retronym** [comment]
We should give this a whirl through the GenBCode optimized Community build, however that Jenkins instance appears to be down.

**Comment by lrytz** [comment]
Started a build here - hope this works as intended: https://jenkins-dbuild.typesafe.com:8499/job/Community-2.11.x-manual/83/parameters/

**Comment by lrytz** [comment]
The community build passed (with `scalac_opts="-Ybackend:GenBCode -Ydelambdafy:method -Yopt:l:method"`)

**Comment by lrytz** [comment]
Any volunteers for reviewing? cc @retronym @gkossakowski @dragos :)

File: src/compiler/scala/tools/nsc/backend/jvm/opt/BytecodeUtils.scala [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    if (method.tryCatchBlocks != null) {
+      method.tryCatchBlocks.iterator().asScala.foreach(l => { add(l.start, l); add(l.handler, l); add(l.end, l) })
+    }
+
+    res.toMap
+  }
+
+  def substituteLabel(reference: AnyRef, from: LabelNode, to: LabelNode): Unit = {
+    def substList(list: java.util.List[LabelNode]) = {
+      list.asScala.zipWithIndex.foreach { case (l, i) =>
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/BytecodeUtils.scala]: `foreachWithIndex(list.asScala)(...)` is more efficient.

File: src/compiler/scala/tools/nsc/backend/jvm/opt/LocalOpt.scala [hunk 2]
```diff
@@ -135,35 +216,107 @@ object LocalOpt {
...
+    method.localVariables.size != initialNumVars
+  }
+
+  /**
+   * The number of local varialbe slots used for parameters and for the `this` reference.
+   */
+  private def parametersSize(method: MethodNode): Int = {
+    // Double / long fields occupy two slots, so we sum up the sizes. Since getSize returns 0 for
+    // void, we have to add `max 1`.
+    val paramsSize = scala.tools.asm.Type.getArgumentTypes(method.desc).map(_.getSize max 1).sum
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/LocalOpt.scala]: Let's avoid the intermediate array creation:

``` scala
getArgumentTypes(method.desc).iterator.map(_.getSize max 1).sum
```

File: src/compiler/scala/tools/nsc/backend/jvm/opt/LocalOpt.scala [hunk 3]
```diff
@@ -135,35 +216,107 @@ object LocalOpt {
...
+    // Add the index of the local variable used by `varIns` to the `renumber` array.
+    def addVar(varIns: VarInsnNode): Unit = {
+      val index = varIns.`var`
+      val isWide = (varIns.getOpcode: @switch) match {
+        case Opcodes.LLOAD | Opcodes.DLOAD | Opcodes.LSTORE | Opcodes.DSTORE => true
+        case _ => false
+      }
+
+      // Ensure the length of `renumber`. Unused variable indices are mapped to -1.
+      val minLenght = if (isWide) index + 2 else index + 1
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/LocalOpt.scala]: typo in `minLenght`

File: src/compiler/scala/tools/nsc/backend/jvm/opt/LocalOpt.scala [hunk 4]
```diff
@@ -135,35 +216,107 @@ object LocalOpt {
...
+      for (i <- renumber.length until minLenght) renumber += -1
+
+      renumber(index) = index
+      if (isWide) renumber(index + 1) = index
+    }
+
+    // first phase: collect all used local variables. if the variable at index x is used, set
+    // renumber(x) = x, otherwsie renumber(x) = -1. if the variable is wide (long or doulbe), set
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/LocalOpt.scala]: "otherwsie", "doulbe"

File: src/compiler/scala/tools/nsc/backend/jvm/opt/LocalOpt.scala [hunk 5]
```diff
@@ -187,4 +340,224 @@ object LocalOpt {
...
+    method.instructions.size != initialSize
+  }
+
+  /**
+   * Apply various simplifications to branching instructions.
+   */
+  def simplifyJumps(method: MethodNode): Boolean = {
+    var changed = false
+
+    val allHanlders = method.tryCatchBlocks.asScala.toSet
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/LocalOpt.scala]: typo "allHanlders"

File: src/compiler/scala/tools/nsc/backend/jvm/opt/LocalOpt.scala [hunk 6]
```diff
@@ -187,4 +340,224 @@ object LocalOpt {
...
+   * rewrite might change the behavior. For xRETURN, the reason is that return insructions may throw
+   * an IllegalMonitorStateException, as described here:
+   *   http://docs.oracle.com/javase/specs/jvms/se8/html/jvms-6.html#jvms-6.5.return
+   */
+  private def simplifyGotoReturn(method: MethodNode, instruction: AbstractInsnNode, inTryBlock: Boolean): Boolean = !inTryBlock && (instruction match {
+    case Goto(jump) =>
+      nextExecutableInstruction(jump.label) match {
+        case Some(target) =>
+          val op = target.getOpcode
+          if ((op >= Opcodes.IRETURN && op <= Opcodes.RETURN) || op == Opcodes.ATHROW) {
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/LocalOpt.scala]: Factor this out to a `def isReturnOp(op: OpCode) = op >= Opcodes.IRETURN && op <= Opcodes.RETURN`

**Comment by retronym** [comment]
LGTM other than those minor cleanups.

File: src/compiler/scala/tools/nsc/backend/jvm/opt/BytecodeUtils.scala (hunk 1 above)
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/BytecodeUtils.scala]: Hmm, there's no global or symbol table in scope, so I can't get to `foreachWithIndex`..

Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/BytecodeUtils.scala]: Ah, just found the `object Collections extends Collections`

Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/opt/BytecodeUtils.scala]: Hmm, but it requires a `toList`, so probably not much faster in the end.. `Collections.foreachWithIndex(list.asScala.toList) { case (l, i) =>`

**Comment by lrytz** [comment]
Last commit addresses the comments

**Pull request merged**